### ISSUE TITLE
Datadog agent pod deployed via stable chart crashing in EKS

### ISSUE THREAD

**Issue by echoboomer** [DataDog/datadog-agent#3114]
**Describe what happened:**
After installing the datadog agent on each node in a Kubernetes cluster using the latest stable chart version via a DaemonSet, one of the agent pods stopped working, resulting in a restart loop of approximately `2482` restarts before manually killing the pod to get it to work again.

Output from failed pods:
```
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 01-check-apikey.sh: executing...
[cont-init.d] 01-check-apikey.sh: exited 0.
[cont-init.d] 50-cri.sh: executing...
[cont-init.d] 50-cri.sh: exited 0.
[cont-init.d] 50-ecs.sh: executing...
[cont-init.d] 50-ecs.sh: exited 0.
[cont-init.d] 50-kubernetes.sh: executing...
[cont-init.d] 50-kubernetes.sh: exited 0.
[cont-init.d] 50-mesos.sh: executing...
[cont-init.d] 50-mesos.sh: exited 0.
[cont-init.d] 51-docker.sh: executing...
[cont-init.d] 51-docker.sh: exited 0.
[cont-init.d] 59-defaults.sh: executing...
[cont-init.d] 59-defaults.sh: exited 0.
[cont-init.d] 60-network-check.sh: executing...
[cont-init.d] 60-network-check.sh: exited 0.
[cont-init.d] 89-copy-customfiles.sh: executing...
[cont-init.d] 89-copy-customfiles.sh: exited 0.
[cont-init.d] done.
[services.d] starting services
[PROCESS] starting process-agent
[NETWORK] starting network-tracer
[ TRACE ] starting trace-agent
[ AGENT ] starting agent
[services.d] done.
[ TRACE ] 2019-03-04 13:23:13 INFO (config.go:214) - Loaded configuration: /etc/datadog-agent/datadog.yaml
[ TRACE ] 2019-03-04 13:23:13 INFO (main.go:35) - received signal 15 (terminated)
[NETWORK] network-tracer exited with code 256, signal 15, restarting in 2 seconds
[PROCESS] process-agent exited with code 256, signal 15, restarting in 2 seconds
[ AGENT ] AGENT EXITED WITH CODE 256, SIGNAL 15, KILLING CONTAINER
[ TRACE ] 2019-03-04 13:23:15 INFO (trace.go:59) - Trace writer initializing with config: {MaxSpansPerPayload:1000 FlushPeriod:5s UpdateInfoPeriod:1m0s SenderConfig:{MaxAge:20m0s MaxQueuedBytes:67108864 MaxQueuedPayloads:-1 ExponentialBackoff:{MaxDuration:2m0s GrowthBase:2 Base:200ms}}}
[ TRACE ] 2019-03-04 13:23:15 INFO (stats.go:47) - Stats writer initializing with config: {MaxEntriesPerPayload:12000 UpdateInfoPeriod:1m0s SenderConfig:{MaxAge:20m0s MaxQueuedBytes:67108864 MaxQueuedPayloads:-1 ExponentialBackoff:{MaxDuration:2m0s GrowthBase:2 Base:200ms}}}
[ TRACE ] 2019-03-04 13:23:15 INFO (service.go:36) - Service writer initializing with config: {FlushPeriod:5s UpdateInfoPeriod:1m0s SenderConfig:{MaxAge:20m0s MaxQueuedBytes:67108864 MaxQueuedPayloads:-1 ExponentialBackoff:{MaxDuration:2m0s GrowthBase:2 Base:200ms}}}
[ TRACE ] 2019-03-04 13:23:15 INFO (main.go:160) - trace-agent running on host i-***
[ TRACE ] 2019-03-04 13:23:15 INFO (api.go:140) - listening for traces at http://0.0.0.0:8126
[ TRACE ] 2019-03-04 13:23:15 INFO (agent.go:140) - exiting
[ TRACE ] 2019-03-04 13:23:15 INFO (concentrator.go:91) - exiting concentrator, computing remaining stats
[ TRACE ] 2019-03-04 13:23:15 INFO (trace.go:134) - exiting trace writer, flushing all remaining traces
[ TRACE ] 2019-03-04 13:23:15 INFO (trace.go:137) - Flushed. Exiting
[ TRACE ] 2019-03-04 13:23:15 INFO (payload.go:179) - exiting payload sender, try flushing whatever is left
[ TRACE ] 2019-03-04 13:23:15 INFO (stats.go:86) - exiting stats writer
[ TRACE ] 2019-03-04 13:23:15 INFO (payload.go:179) - exiting payload sender, try flushing whatever is left
[ TRACE ] 2019-03-04 13:23:15 INFO (service.go:107) - exiting service writer, flushing all modified services
[ TRACE ] 2019-03-04 13:23:15 INFO (payload.go:179) - exiting payload sender, try flushing whatever is left
[ TRACE ] trace-agent exited with code 0, disabling
```
```
datadog-dwz4w   0/1     CrashLoopBackOff   2482       9d
```
Chart values file:
```
daemonset:
  updateStrategy: RollingUpdate

datadog:
  apiKey: ${apiKey}
  apmEnabled: true
  appKey: ${appKey}
  collectEvents: true
  leaderElection: true
  logsConfigContainerCollectAll: true
  logsEnabled: true
  processAgentEnabled: true
  site: datadoghq.com
  podLabelsAsTags:
    [redacted]

rbac:
  create: true
```
**Describe what you expected:**
For the pod to actually be killed so it would start up properly which worked when done manually. At a minimum, I'd like to figure out what caused the error so we can avoid it elsewhere.

**Steps to reproduce the issue:**
This has only happened once and I haven't been able to reproduce it.

**Additional environment details (Operating System, Cloud provider, etc):**
AWS EKS, `v1.11.5-eks-6bad6d`

**Comment by grahampfau** [comment]
We're seeing this too in multiple EKS clusters, where one pod in constantly in a bad state.

One thing I've noticed is that the unhealthy pod is always on the node where `datadog-kube-state-metrics` is scheduled.

**Comment by guizmaii** [comment]
@grahampfau You should increase:

```yaml
  resources:
    requests:
      memory: 256Mi
    limits:
      memory: <here>
```

**Comment by grahampfau** [comment]
@guizmaii That's for the suggestion, but I'm really hesitant to set that limit for two reasons:

- We'd lose the `qosClass: Guaranteed`
- It's only one pod in the whole DaemonSet that's hitting this problem, and it's always on the node running `datadog-kube-state-metrics`, so I think there's a larger problem to be fixed.

**Comment by grahampfau** [comment]
@echoboomer We worked with DataDog on this, and the issue for us was that kube-state-metrics was collecting so much that it was causing the DataDog agent pod colocated with ksm to hit its resource limits, because that's the only pod that collects those particular metrics.

The workaround was to run an agent separate from the daemonset as a sidecar to ksm.

Here's the gist I worked off of to get that sidecar: https://gist.github.com/gzysk8/a829420d5ee9e9adff29c81f42248133

**Comment by echoboomer** [comment]
@grahampfau we're deploying the official Datadog Helm chart using Terraform, so I don't see any easy way to add the sidecar manifest in without managing our own local version of the chart. Is that what you're doing?

**Comment by grahampfau** [comment]
@echoboomer Essentially. We're installing the stable chart (with `kubeStateMetrics.enabled=true` and `kube-state-metrics.rbac.create=true` so we get the RBAC), deleting the ksm deployment that the chart made and then applying a modified version of that yaml on top to make the ksm with agent sidecar.

We're using a script instead of Terraform though for helm charts (although that may change).

So yeah, the fix for this is as kludgy as it sounds :stuck_out_tongue:

**Comment by CharlyF** [comment]
Hello all,

Thank you very much for sharing your solutions. This is an issue we have recently addressed, and while we are still working on advertising it I figured it'd be worthwhile sharing it here.

The solution is to leverage Cluster Level Checks. This is a new feature that allows to monitor resources that can't be "autodiscovered" (aka pods). Typically, if you want to run a Postgres check against an RDS database or an HTTP check against a LB. It can also be used to schedule checks on a dedicated instance part of a pool of runners.

You can read more about this in the official doc:
https://docs.datadoghq.com/agent/autodiscovery/clusterchecks/

tl;dr: Using the [Cluster Agent](https://docs.datadoghq.com/agent/kubernetes/cluster/) as a scheduler, you pass it files to schedule the checks but you can also annotate your services for the Cluster Agent to schedule an associated check on one agent.

This is typically how we monitor Kube-State-Metrics internally. We have created a standalone deployment that we call Cluster Check Runner (or Worker). Its role will be to run the Cluster Level Checks scheduled by the Datadog Cluster Agent. The benefit of using a dedicated worker is so you can bump it's memory/cpu limits and requirements. This is to address memory intensive checks such as KSM.

We have enabled those configurations in our Helm Chart as well, so you should be able to make it all work out of the box. You can look at the deployment file of the Cluster Check Worker [here](https://github.com/helm/charts/blob/master/stable/datadog/templates/agent-clusterchecks-deployment.yaml).
I just realized that this is not well documented, so I am going to make sure we update this.
The variables to use are `clusterAgent.clusterChecks.enabled` and `clusterchecksDeployment.*`

For instance, you could split the KSM deployments by collector and have:
For the Pod deployment:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: kube-state-metrics-pods
[...]
      containers:
      - name: kube-state-metrics
        image: quay.io/kube-state-metrics:1.5.0
        command:
          - /kube-state-metrics
          - --telemetry-port=8081
          - --port=8080
          - --collectors=pods
```
Associated with the following annotated service:
```
apiVersion: v1
kind: Service
metadata:
  name: kube-state-metrics-pods
  namespace: kube-monitoring
  annotations:
    ad.datadoghq.com/service.check_names: '["kubernetes_state"]'
    ad.datadoghq.com/service.init_configs: '[{}]'
    ad.datadoghq.com/service.instances: |
      [
        {
          "kube_state_url": "http://%%host%%:8080/metrics",
          "prometheus_timeout": 30,
          "min_collection_interval": 30,
          "send_pod_phase_service_checks": false,
        }
      ]
spec:
  selector:
    app: kube-state-metrics-pods
  ports:
  - port: 8080
    protocol: TCP
    targetPort: 8080
```

This would result in:
```
root@datadog-cluster-agent-XXX:/# agent clusterchecks
=== 3 node-agents reporting ===

Name                                                 Running checks
NodeA                                                        2
NodeB                                                        1
NodeC                                                        1
[...]
=== kubernetes_state check ===
Source: kubernetes-services
Instance ID: kubernetes_state:3a0e4XXXXX659fb
empty_default_hostname: true
kube_state_url: http://10.17.111.222:8080/metrics
min_collection_interval: 30
prometheus_timeout: 30
send_pod_phase_service_checks: false
tags:
- kubernetes_cluster:my-cluster
- kube_service:kube-state-metrics-pods
- kube_namespace:kube-monitoring
~
Init Config:
{}
===
```
That will be running on the Cluster Check Runner:
```
root@datadog-agent-cluster-worker-XXX-zj5mf:/# agent status
Getting the status from the agent.
[...]
    kubernetes_state (4.2.0)
    ------------------------
      Instance ID: kubernetes_state:3a0e4XXXXX659fb [OK]
      Total Runs: 219
      Metric Samples: Last Run: 243,319, Total: 1.2 M
[...]
```

I think this solution could help many of you, I would also love to gather feedback!

Best,
.C

**Comment by echoboomer** [comment]
Thanks, @CharlyF - this is really good info. I've already got a PR open to trial this out and, given the occurrence frequency of these issues, I don't think it's logical to need to report back in on this thread, but I can totally give an update once it has rolled out for my situation.

**Comment by CharlyF** [comment]
Awesome. Whatever can benefit the community, feel free to post here. if you have questions or for anything else feel free to shoot me an email: c@datadoghq.com. I'd be happy to redirect to the right folks internally.

**Comment by echoboomer** [comment]
I'm going to go ahead and close this. Since enabling `clusterChecks`, we really haven't seen any of the same issues. Thanks again for all of the help.
